Granular Date Logic

laambrosino opened the pull request:
Currently working on slightly altering this functionality in synthea to work with the way the module builder is saving date logic JSON, so this side of things seems to be working correctly but the backend side isn't yet.

arscan commented:
It appears that you can add in an invalid date.  Should we try to parse the date and if it doesn't parse, reject the change (just revert it to what it was before?).  For example, if the date is 2012-05-06, and the user changes it to 2012-aa-06, maybe just have it revert back to 2012-05-06?

An alternate solution would be to add a warning in the 'analysis' portion, which would also catch errors for people doing manual editing.  But I think it would be easier to ensure dates are valid at input time.

I don't think we should allow the module builder to output data that is obviously wrong.  I'm ok with subtly wrong (if hard to identify), but this seems like it should be relatively easy to prevent.

laambrosino commented:
I changed the inputs for all of the date aspects from `RIEInput` to `RIENumber` so now if you try to input anything other than a number it will just revert back to what was there before. I also added some warnings if an input doesn't make sense, for example entering 100 for the month. I wasn't sure if we'd want to make a range for the year input (show a warning for a year 500000, etc.) but for now I only made a warning if `year < 1`
